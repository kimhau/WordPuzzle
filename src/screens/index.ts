export { default as Congratz } from './Congratz/Congratz';
export { default as Game } from './Game/Game';
export { default as Home } from './Home/Home';
export { default as Leaderboard } from './Leaderboard/Leaderboard';
export { default as Startup } from './Startup/Startup';
