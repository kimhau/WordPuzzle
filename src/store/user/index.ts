import { createSlice } from '@reduxjs/toolkit';

const slice = createSlice({
  name: 'user',
  initialState: { name: null, points: 0 } as UserState,
  reducers: {
    updateUser: (state, { payload: { name, points } }: UserPayload) => {
      if (typeof name !== 'undefined') {
        state.name = name;
      }
      if (typeof points !== 'undefined') {
        state.points = points;
      }
    },
    updatePoints: (state, { payload: { points } }: UserPayload) => {
      if (typeof points !== 'undefined') {
        state.points += points;
      }
    },
  },
});

export const { updateUser, updatePoints } = slice.actions;

export default slice.reducer;

export type UserState = {
  name: string | null;
  points: number;
};

export type UserPayload = {
  payload: Partial<UserState>;
};
