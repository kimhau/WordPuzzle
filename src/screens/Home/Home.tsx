import React, { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { SafeAreaView, Text, View } from 'react-native';
import { TouchableOpacity } from 'react-native-gesture-handler';
import { useTheme } from 'WordPuzzle/src/hooks';
import { ApplicationScreenProps } from '../../../@types/navigation';

const Home = ({ navigation }: ApplicationScreenProps) => {
  const { t } = useTranslation('home');
  const { Fonts, Common, Layout, Gutters } = useTheme();
  const [selectedCategory, setSelectedCategory] = useState<string>();

  const handleCategorySelect = (category: string) => {
    setSelectedCategory(category);
  };

  const handleStartPress = () => {
    // Navigate to game screen with selectedCategory
    console.log(`Starting game with category: ${selectedCategory}`);
    navigation.navigate({
      name: 'Game',
      params: {
        category: selectedCategory,
      },
    });
  };

  const handleLeaderboardPress = () => {
    // Navigate to leaderboard screen
    navigation.navigate({
      name: 'Leaderboard',
    });
  };

  return (
    <SafeAreaView
      style={[Layout.fill, Gutters.regularPadding, Layout.justifyContentAround]}
    >
      <Text style={[Fonts.titleRegular, Fonts.textCenter]}>
        {t('wordPuzzle')}
      </Text>
      <View style={[Layout.column, Layout.gap, Gutters.smallPadding]}>
        <TouchableOpacity
          style={[
            Common.button.rounded,
            {
              backgroundColor:
                selectedCategory === 'animals' ? 'orange' : 'gray',
            },
          ]}
          onPress={() => handleCategorySelect('animals')}
        >
          <Text style={Fonts.textCenter}>{t('animals')}</Text>
        </TouchableOpacity>
        <TouchableOpacity
          style={[
            Common.button.rounded,
            {
              backgroundColor:
                selectedCategory === 'countries' ? 'orange' : 'gray',
            },
          ]}
          onPress={() => handleCategorySelect('countries')}
        >
          <Text style={Fonts.textCenter}>{t('countries')}</Text>
        </TouchableOpacity>
        <TouchableOpacity
          style={[
            Common.button.rounded,
            {
              backgroundColor:
                selectedCategory === 'chemicals' ? 'orange' : 'gray',
            },
          ]}
          onPress={() => handleCategorySelect('chemicals')}
        >
          <Text style={Fonts.textCenter}>{t('chemicals')}</Text>
        </TouchableOpacity>
      </View>
      <TouchableOpacity
        testID="start-button"
        disabled={!selectedCategory}
        style={[Common.button.rounded]}
        onPress={handleStartPress}
      >
        <Text style={Fonts.textBold}>{t('start')}</Text>
      </TouchableOpacity>
      <TouchableOpacity onPress={handleLeaderboardPress}>
        <Text style={[Fonts.textRight, Fonts.textPrimary]}>
          {t('go-to-leaderboard')}
        </Text>
      </TouchableOpacity>
    </SafeAreaView>
  );
};

export default Home;
