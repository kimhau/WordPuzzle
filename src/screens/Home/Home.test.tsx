import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';
import { Provider } from 'react-redux';
import { store } from 'WordPuzzle/src/store';
import Home from './Home';

describe('Home', () => {
  it('renders correctly', () => {
    const { getByText, toJSON, getByTestId } = render(
      <Provider store={store}>
        <Home />
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
    expect(getByText('wordPuzzle')).toBeDefined();
    expect(getByText('animals')).toBeDefined();
    expect(getByText('countries')).toBeDefined();
    expect(getByText('chemicals')).toBeDefined();
    expect(getByTestId('start-button').props.enabled).toBe(false);
    expect(getByText('go-to-leaderboard')).toBeDefined();
  });

  it('selects a category', () => {
    const { getByText, getByTestId } = render(
      <Provider store={store}>
        <Home />
      </Provider>,
    );
    const animalsButton = getByText('animals');
    const countriesButton = getByText('countries');
    const chemicalsButton = getByText('chemicals');

    fireEvent.press(animalsButton);
    expect(animalsButton.parent?.parent?.props.style.backgroundColor).toBe(
      'orange',
    );
    expect(countriesButton.parent?.parent?.props.style.backgroundColor).toBe(
      'gray',
    );
    expect(chemicalsButton.parent?.parent?.props.style.backgroundColor).toBe(
      'gray',
    );
    expect(getByTestId('start-button').props.enabled).toBe(true);

    fireEvent.press(countriesButton);
    expect(animalsButton.parent?.parent?.props.style.backgroundColor).toBe(
      'gray',
    );
    expect(countriesButton.parent?.parent?.props.style.backgroundColor).toBe(
      'orange',
    );
    expect(chemicalsButton.parent?.parent?.props.style.backgroundColor).toBe(
      'gray',
    );
    expect(getByTestId('start-button').props.enabled).toBe(true);

    fireEvent.press(chemicalsButton);
    expect(animalsButton.parent?.parent?.props.style.backgroundColor).toBe(
      'gray',
    );
    expect(countriesButton.parent?.parent?.props.style.backgroundColor).toBe(
      'gray',
    );
    expect(chemicalsButton.parent?.parent?.props.style.backgroundColor).toBe(
      'orange',
    );
    expect(getByTestId('start-button').props.enabled).toBe(true);
  });

  it('navigates to game screen with selected category', () => {
    const navigation = { navigate: jest.fn() };
    const { getByText, getByTestId } = render(
      <Provider store={store}>
        <Home navigation={navigation} />
      </Provider>,
    );
    const animalsButton = getByText('animals');
    fireEvent.press(animalsButton);

    const startButton = getByTestId('start-button');
    fireEvent.press(startButton);
    expect(navigation.navigate).toHaveBeenCalledWith({
      name: 'Game',
      params: { category: 'animals' },
    });
  });

  it('navigates to leaderboard screen', () => {
    const navigation = { navigate: jest.fn() };
    const { getByText } = render(
      <Provider store={store}>
        <Home navigation={navigation} />
      </Provider>,
    );
    const leaderboardButton = getByText('go-to-leaderboard');
    fireEvent.press(leaderboardButton);
    expect(navigation.navigate).toHaveBeenCalledWith({ name: 'Leaderboard' });
  });
});
