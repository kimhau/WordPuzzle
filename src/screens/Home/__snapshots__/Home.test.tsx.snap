// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders correctly 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "flex": 1,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "flex": 1,
        },
        {
          "padding": 30,
        },
        {
          "justifyContent": "space-around",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#212529",
            "fontSize": 40,
            "fontWeight": "bold",
          },
          {
            "textAlign": "center",
          },
        ]
      }
    >
      wordPuzzle
    </Text>
    <View
      style={
        [
          {
            "flexDirection": "column",
          },
          {
            "gap": 8,
          },
          {
            "padding": 20,
          },
        ]
      }
    >
      <RNGestureHandlerButton
        collapsable={false}
        delayLongPress={600}
        enabled={true}
        exclusive={true}
        handlerTag={1}
        handlerType="NativeViewGestureHandler"
        onGestureEvent={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onHandlerStateChange={[Function]}
        rippleColor={0}
        touchSoundDisabled={false}
      >
        <View
          accessible={true}
          collapsable={false}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "gray",
              "borderRadius": 10,
              "height": 40,
              "justifyContent": "center",
              "opacity": 1,
              "paddingHorizontal": 30,
            }
          }
        >
          <Text
            style={
              {
                "textAlign": "center",
              }
            }
          >
            animals
          </Text>
        </View>
      </RNGestureHandlerButton>
      <RNGestureHandlerButton
        collapsable={false}
        delayLongPress={600}
        enabled={true}
        exclusive={true}
        handlerTag={2}
        handlerType="NativeViewGestureHandler"
        onGestureEvent={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onHandlerStateChange={[Function]}
        rippleColor={0}
        touchSoundDisabled={false}
      >
        <View
          accessible={true}
          collapsable={false}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "gray",
              "borderRadius": 10,
              "height": 40,
              "justifyContent": "center",
              "opacity": 1,
              "paddingHorizontal": 30,
            }
          }
        >
          <Text
            style={
              {
                "textAlign": "center",
              }
            }
          >
            countries
          </Text>
        </View>
      </RNGestureHandlerButton>
      <RNGestureHandlerButton
        collapsable={false}
        delayLongPress={600}
        enabled={true}
        exclusive={true}
        handlerTag={3}
        handlerType="NativeViewGestureHandler"
        onGestureEvent={[Function]}
        onGestureHandlerEvent={[Function]}
        onGestureHandlerStateChange={[Function]}
        onHandlerStateChange={[Function]}
        rippleColor={0}
        touchSoundDisabled={false}
      >
        <View
          accessible={true}
          collapsable={false}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "gray",
              "borderRadius": 10,
              "height": 40,
              "justifyContent": "center",
              "opacity": 1,
              "paddingHorizontal": 30,
            }
          }
        >
          <Text
            style={
              {
                "textAlign": "center",
              }
            }
          >
            chemicals
          </Text>
        </View>
      </RNGestureHandlerButton>
    </View>
    <RNGestureHandlerButton
      collapsable={false}
      delayLongPress={600}
      enabled={false}
      exclusive={true}
      handlerTag={4}
      handlerType="NativeViewGestureHandler"
      onGestureEvent={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onHandlerStateChange={[Function]}
      rippleColor={0}
      testID="start-button"
      touchSoundDisabled={false}
    >
      <View
        accessible={true}
        collapsable={false}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#E14032",
            "borderRadius": 10,
            "height": 40,
            "justifyContent": "center",
            "opacity": 1,
            "paddingHorizontal": 30,
          }
        }
      >
        <Text
          style={
            {
              "fontWeight": "bold",
            }
          }
        >
          start
        </Text>
      </View>
    </RNGestureHandlerButton>
    <RNGestureHandlerButton
      collapsable={false}
      delayLongPress={600}
      enabled={true}
      exclusive={true}
      handlerTag={5}
      handlerType="NativeViewGestureHandler"
      onGestureEvent={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onHandlerStateChange={[Function]}
      rippleColor={0}
      touchSoundDisabled={false}
    >
      <View
        accessible={true}
        collapsable={false}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "right",
              },
              {
                "color": "#E14032",
              },
            ]
          }
        >
          go-to-leaderboard
        </Text>
      </View>
    </RNGestureHandlerButton>
  </View>
</RCTSafeAreaView>
`;
